proposals/index-range-changes.md

```diff
@@ -0,0 +1,297 @@
+    private int[] _array = new[] { 1, 2, 3 };
+
+    int Length {
+        get {
+            Console.Write("Length ");
+            return _array.Length;
+        }
+    }
+
+    int[] Slice(int start, int length) { 
+        var slice = new int[length];
+        Array.Copy(_array, start, slice, 0, length);
+        return slice;
+    }
+}
+
+class SideEffect {
+    Collection Get() {
+        Console.Write("Get ");
+        return new Collection();
+    }
+
+    void Use() { 
+        var array = Get()[0..2];
+        Console.WriteLine(array.length);
+    }
+}
+```
+
+This code will print "Get Length 2". 
+
+The language will special case the following known types: 
+
+- `string`: the method `Substring` will be used instead of `Slice`.
+- `array`: the method `System.Reflection.CompilerServices.GetSubArray` will be used instead of `Slice`.
+
+## Open Issues
+
+## Considerations
+
+### Detect Indexable based on ICollection
+The inspiration for this proposal was collection initializers. Using the structure of a type to convey that it had
+opted into a feature. In the case of collection initializers types can opt into the feature by implementing the 
+interface `IEnumerable` (non generic).
+
+This proposal initially required that types implement `ICollection` in order to qualify as Indexable. That required a number of special cases though:
+required a number of special cases:
+
+- `ref struct`: these cannot implement interfaces yet types like `Span<T>` are ideal for index / range support. 
+- `string`: does not implement `ICollection` and adding that `interface` has a large cost.
+
+This means to support key types special casing is already needed. The special casing of `string` is less interesting 
+as the language does this in other areas (`foreach` lowering, constants, etc ...). The special casing of `ref struct`
+is more concerning as it's special casing an entire class of types. They get labeled as Indexable if they simply have
+a property named `Count` with a return type of `int`. 
+
+After consideration the design was normalized to say that any type which has a property `Count` / `Length` with a 
+return type of `int` is Indexable. That removes all special casing, even for `string` and arrays.
+
+### Detect just Count
+Detecting on the property names `Count` or `Length` does complicate the design a bit. Picking just one to standardize
+though is not sufficient as it ends up excluding a large number of types:
+
+- Use `Length`: excludes pretty much every collection in System.Collections and sub-namespaces. Those tend to derive 
+from `ICollection` and hence prefer `Count` over length.
+- Use `Count`: excludes `string`, arrays, `Span<T>` and most `ref struct` based types
+
+The extra complication on the initial detection of Indexable types is outweighed by its simplification in other 
+aspects.
+
+### Choice of Slice as a anme
+The name `Slice` was chosen as it's the de-facto standard name for slice style operations in .NET. Starting with 
```

Comment:
Will the compiler special case the string here as it doesn't have Slice but it has Substring?

Reply:
@tarekgh Yup!  that is called out explicitly elsewhere in the doc.

Reply:
The following statement in the document could be confusing then:

```
After consideration the design was normalized to say that any type which has a property `Count` / `Length` with a 
return type of `int` is Indexable. That removes all special casing, even for `string` and arrays.
```